File: java/believe/gamestate/levelstate/LevelState.java
package believe.gamestate.levelstate;

import believe.action.ChangeToTemporaryStateAction;
import believe.character.Character.DamageListener;
import believe.character.Faction;
import believe.character.playable.PlayableCharacter;
import believe.character.playable.PlayableCharacterFactory;
import believe.core.io.FontLoader;
import believe.datamodel.MutableValue;
import believe.gamestate.GameStateBase;
import believe.gamestate.temporarystate.GameOverState;
import believe.gamestate.temporarystate.GamePausedOverlay;
import believe.gamestate.temporarystate.OverlayablePrecedingState;
import believe.gamestate.temporarystate.PrecedingState;
import believe.gui.ProgressBar;
import believe.level.LevelData;
import believe.level.LevelManager;
import believe.map.data.MapData;
import believe.map.gui.PlayArea;
import believe.physics.manager.PhysicsManager;
import javax.annotation.Nullable;
import org.newdawn.slick.GameContainer;
import org.newdawn.slick.Graphics;
import org.newdawn.slick.Image;
import org.newdawn.slick.Input;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.state.StateBasedGame;

import java.util.Optional;
import java.util.stream.Stream;

public abstract class LevelState extends GameStateBase
    implements OverlayablePrecedingState, DamageListener {

  private final LevelManager levelManager;
  private final GameContainer container;
  private final ChangeToTemporaryStateAction<OverlayablePrecedingState> pauseAction;
  private final ChangeToTemporaryStateAction<PrecedingState> gameOverAction;
  private final PlayableCharacterFactory playableCharacterFactory;
  private final MutableValue<Optional<PlayableCharacter>> currentPlayableCharacter;
  private final PhysicsManager physicsManager;

  private boolean enteringFromPauseMenu;
  @Nullable private LevelData levelData;
  private ProgressBar focusBar;

  protected StateBasedGame game;
  @Nullable protected PlayArea playArea;
  @Nullable protected PlayableCharacter player;

  public LevelState(
      GameContainer container,
      StateBasedGame game,
      LevelManager levelManager,
      PhysicsManager physicsManager,
      FontLoader fontLoader,
      PlayableCharacterFactory playableCharacterFactory,
      MutableValue<Optional<PlayableCharacter>> currentPlayableCharacter) {
    this.container = container;
    this.game = game;
    this.levelManager = levelManager;
    this.physicsManager = physicsManager;
    this.pauseAction = new ChangeToTemporaryStateAction<>(GamePausedOverlay.class, this, game);
    this.gameOverAction = new ChangeToTemporaryStateAction<>(GameOverState.class, this, game);
    this.playableCharacterFactory = playableCharacterFactory;
    this.currentPlayableCharacter = currentPlayableCharacter;
    this.focusBar = new ProgressBar(container, fontLoader.getBaseFontAtSize(15));
    this.focusBar.setBorderSize(1);
    this.focusBar.setTextPadding(0);
  }

  @Override
  public void init(GameContainer container, StateBasedGame game) throws SlickException {}

  @Override
  public void render(GameContainer container, StateBasedGame game, Graphics g)
      throws SlickException {
    playArea.render(container, g);
  }

  @Override
  public void update(GameContainer container, StateBasedGame game, int delta)
      throws SlickException {
    playArea.update(delta);
    player.update(delta);
    physicsManager.update(delta);
    focusBar.setProgress(player.getFocus());
  }

  @Override
  public void keyPressed(int key, char c) {
    super.keyPressed(key, c);

    switch (key) {
      case Input.KEY_ESCAPE:
        enteringFromPauseMenu = true;
        pauseAction.activate();
        break;
    }
  }

  @Override
  public void reset() {
    player.setLocation(
        levelData.getMapData().tiledMapData().playerStartX(),
        levelData.getMapData().tiledMapData().playerStartY() - player.getHeight());
    player.setVerticalSpeed(0);
    player.heal(1f);
    playArea.resetLayout();
    initPhysics();
  }

  @Override
  public void enter(GameContainer container, StateBasedGame game) throws SlickException {
    super.enter(container, game);
    if (!enteringFromPauseMenu) {
      levelData = getLevelData();
      MapData mapData = levelData.getMapData();
      player =
          playableCharacterFactory.create(
              this,
              isOnRails(),
              mapData.tiledMapData().playerStartX(),
              mapData.tiledMapData().playerStartY());
      currentPlayableCharacter.update(Optional.of(player));
      playArea = providePlayArea(mapData, player);
      focusBar.setText("Focus");
      playArea.addHudChild(focusBar, 0.02f, 0.05f, 0.15f, 0.07f);
      initPhysics();

      levelEnter(container, game);
      levelData.getInitialCommand().execute();
    }
  }

  private LevelData getLevelData() {
    LevelData levelData = levelManager.getLevel(getLevelName());
    if (levelData == null) {
      // TODO(#16): Show an error screen instead of throwing an exception.
      throw new IllegalStateException("Cannot initialize level due to missing level data.");
    }
    return levelData;
  }

  public void reloadLevel() {
    int playerX = player.getX();
    int playerY = player.getY();
    levelData = getLevelData();
    playArea.reloadMap(levelData.getMapData());
    reset();
    player.setLocation(playerX, playerY);
  }

  @Override
  public void exitingFollowingState() {
    this.enteringFromPauseMenu = false;
  }

  private void initPhysics() {
    physicsManager.reset();
    MapData mapData = levelData.getMapData();
    Stream.concat(
            mapData.tiledMapData().objectLayers().stream()
                .flatMap(
                    objectLayerData ->
                        objectLayerData.generatedMapEntityData().physicsManageables().stream()),
            mapData.tiledMapData().layers().stream()
                .flatMap(
                    layerData -> layerData.generatedMapEntityData().physicsManageables().stream()))
        .forEach(entity -> entity.addToPhysicsManager(physicsManager));
    player.addToPhysicsManager(physicsManager);
  }

  @Override
  public Image getCurrentScreenshot() throws SlickException {
    Image screenshot = new Image(container.getWidth(), container.getHeight());
    container.getGraphics().copyArea(screenshot, 0, 0);
    return screenshot;
  }

  protected abstract boolean isOnRails();

  protected abstract String getLevelName();

  protected abstract String getMusicLocation();

  protected abstract PlayArea providePlayArea(MapData mapData, PlayableCharacter player);

  protected abstract void levelEnter(GameContainer container, StateBasedGame game)
      throws SlickException;

  @Override
  public void damageInflicted(float currentFocus, Faction inflictor) {
    if (currentFocus <= 0 && inflictor != Faction.NONE && inflictor != player.getFaction()) {
      gameOverAction.activate();
    }
  }
}

File: java/believe/gamestate/levelstate/LevelState.kt
package believe.gamestate.levelstate

import believe.action.ChangeToTemporaryStateAction
import believe.character.Character.DamageListener
import believe.character.Faction
import believe.character.playable.PlayableCharacter
import believe.character.playable.PlayableCharacterFactory
import believe.core.io.FontLoader
import believe.datamodel.MutableValue
import believe.gamestate.GameStateBase
import believe.gamestate.temporarystate.GameOverState
import believe.gamestate.temporarystate.GamePausedOverlay
import believe.gamestate.temporarystate.OverlayablePrecedingState
import believe.gamestate.temporarystate.PrecedingState
import believe.gui.ProgressBar
import believe.level.LevelData
import believe.level.LevelManager
import believe.map.data.LayerData
import believe.map.data.MapData
import believe.map.data.ObjectLayerData
import believe.map.gui.PlayArea
import believe.physics.manager.PhysicsManageable
import believe.physics.manager.PhysicsManager
import believe.scene.GeneratedMapEntityData
import org.newdawn.slick.GameContainer
import org.newdawn.slick.Graphics
import org.newdawn.slick.Image
import org.newdawn.slick.Input
import org.newdawn.slick.SlickException
import org.newdawn.slick.state.StateBasedGame
import java.util.*
import java.util.stream.Stream

abstract class LevelState(
    private val container: GameContainer,
    protected var game: StateBasedGame,
    private val levelManager: LevelManager,
    private val physicsManager: PhysicsManager,
    fontLoader: FontLoader,
    private val playableCharacterFactory: PlayableCharacterFactory,
    private val currentPlayableCharacter: MutableValue<Optional<PlayableCharacter>>
) : GameStateBase(), OverlayablePrecedingState, DamageListener {

    private val pauseAction: ChangeToTemporaryStateAction<OverlayablePrecedingState> =
        ChangeToTemporaryStateAction(
            GamePausedOverlay::class.java, this, game
        )
    private val gameOverAction: ChangeToTemporaryStateAction<PrecedingState> =
        ChangeToTemporaryStateAction(GameOverState::class.java, this, game)
    private var enteringFromPauseMenu = false
    private var levelData: LevelData? = null
    private val focusBar: ProgressBar =
        ProgressBar(container, fontLoader.getBaseFontAtSize(15f)).apply {
            setBorderSize(1)
            setTextPadding(0)
        }
    protected var playArea: PlayArea? = null
    protected var player: PlayableCharacter? = null

    protected abstract val isOnRails: Boolean
    protected abstract val levelName: String?
    protected abstract val musicLocation: String?

    @Throws(SlickException::class)
    override fun init(container: GameContainer, game: StateBasedGame) {
    }

    @Throws(SlickException::class)
    override fun render(
        container: GameContainer, game: StateBasedGame, g: Graphics
    ) {
        playArea?.render(container, g)
    }

    @Throws(SlickException::class)
    override fun update(
        container: GameContainer, game: StateBasedGame, delta: Int
    ) {
        playArea?.update(delta)
        physicsManager.update(delta.toLong())
        player?.let {
            it.update(delta.toLong())
            focusBar.setProgress(it.focus)
        }
    }

    override fun keyPressed(key: Int, c: Char) {
        super.keyPressed(key, c)
        when (key) {
            Input.KEY_ESCAPE -> {
                enteringFromPauseMenu = true
                pauseAction.activate()
            }
        }
    }

    override fun reset() {
        levelData?.let {
            player?.setLocation(
                it.mapData.tiledMapData().playerStartX(),
                it.mapData.tiledMapData().playerStartY() - (player?.height ?: 0)
            )
        }
        player?.verticalSpeed = 0f
        player?.heal(1f)
        playArea?.resetLayout()
        initPhysics()
    }

    @Throws(SlickException::class)
    override fun enter(container: GameContainer, game: StateBasedGame) {
        super.enter(container, game)
        if (!enteringFromPauseMenu) {
            levelData = getLevelData()
            levelData?.let {
                val mapData = it.mapData
                player = playableCharacterFactory.create(
                    this,
                    isOnRails,
                    mapData.tiledMapData().playerStartX(),
                    mapData.tiledMapData().playerStartY()
                ).also { playableCharacter ->
                    currentPlayableCharacter.update(Optional.of(playableCharacter))
                }
                playArea = providePlayArea(mapData, player)
            }
            focusBar.setText("Focus")
            playArea?.addHudChild(focusBar, 0.02f, 0.05f, 0.15f, 0.07f)
            initPhysics()
            levelEnter(container, game)
            levelData?.initialCommand?.execute()
        }
    }

    private fun getLevelData(): LevelData {
        return levelName?.let { levelManager.getLevel(it) }
            ?: // TODO(#16): Show an error screen instead of throwing an exception.
            throw IllegalStateException("Cannot initialize level due to missing level data.")
    }

    fun reloadLevel() {
        val playerX = player?.x ?: 0
        val playerY = player?.y ?: 0
        levelData = getLevelData()
        levelData?.let {
            playArea?.reloadMap(it.mapData)
        }
        reset()
        player?.setLocation(playerX, playerY)
    }

    override fun exitingFollowingState() {
        enteringFromPauseMenu = false
    }

    private fun initPhysics() {
        physicsManager.reset()
        val mapData = levelData!!.mapData
        val generatedMapEntityDataBuilder = GeneratedMapEntityData.newBuilder()
        mapData.tiledMapData().objectLayers().stream().forEach { objectLayerData: ObjectLayerData ->
            objectLayerData.objectFactories()
                .forEach { it.createAndAddTo(generatedMapEntityDataBuilder) }
        }

        generatedMapEntityDataBuilder.build().physicsManageables()
            .forEach { entity: PhysicsManageable ->
                entity.addToPhysicsManager(physicsManager)
            }
        player!!.addToPhysicsManager(physicsManager)
    }

    @Throws(SlickException::class)
    override fun getCurrentScreenshot(): Image {
        val screenshot = Image(container.width, container.height)
        container.graphics.copyArea(screenshot, 0, 0)
        return screenshot
    }

    protected abstract fun providePlayArea(mapData: MapData?, player: PlayableCharacter?): PlayArea?

    @Throws(SlickException::class)
    protected abstract fun levelEnter(container: GameContainer?, game: StateBasedGame?)

    override fun damageInflicted(
        currentFocus: Float, inflictor: Faction
    ) {
        if (currentFocus <= 0 && inflictor != Faction.NONE && inflictor != player?.faction) {
            gameOverAction.activate()
        }
    }
}